[OPIK-3012] [BE] Add workspace dashboards CRUD functionality

thiagohora opened the pull request:
## Details

This PR implements the complete CRUD functionality for workspace-level custom dashboards. The implementation includes:

### Core Features
- **Create Dashboard**: Create custom dashboards with name, description, and JSON configuration
- **Read Dashboard**: Get dashboard by ID and list dashboards with pagination
- **Update Dashboard**: Update dashboard fields with optimistic locking (version-based)
- **Delete Dashboard**: Idempotent delete operation (returns 204 regardless of existence)

### Key Components

**Database Layer:**
- New `dashboards` table with UUIDv7 IDs, workspace scoping, and optimistic locking
- JSON column for flexible dashboard configuration storage
- Unique constraints on workspace+name and workspace+slug combinations
- Custom `JsonNodeColumnMapper` for JDBI3 JSON handling

**Domain Layer:**
- `DashboardDAO`: JDBI3 interface for database operations
- `DashboardService`: Business logic including slug generation and validation
- `DashboardConfigValidator`: JSON schema validation for dashboard configurations
- `SlugUtils`: URL-safe slug generation with deduplication

**API Layer:**
- RESTful endpoints under `/v1/private/dashboards`
- Support for pagination, filtering by name, and sorting
- OpenAPI/Swagger documentation
- JsonView annotations for proper field visibility (read-only vs writable)

**Testing:**
- Comprehensive resource tests following `OptimizationsResourceTest` pattern
- `DashboardResourceClient` for test API interactions
- 23 test cases covering all CRUD operations, validation, and edge cases
- Test isolation with unique data generation

### Technical Decisions

1. **Slug Generation**: Automatic URL-safe slug generation from dashboard name with deduplication per workspace
2. **Optimistic Locking**: Version field prevents concurrent update conflicts
3. **Idempotent DELETE**: Returns 204 No Content for both existing and non-existing resources (REST best practice)
4. **JSON Schema Validation**: Validates dashboard config against predefined schema (currently v1)
5. **Workspace Scoping**: All operations are scoped to workspace for proper multi-tenancy

## Change checklist
- [x] User facing
- [ ] Documentation update

## Issues
- Resolves OPIK-3012

## Testing

All 23 tests passing in `DashboardsResourceTest`:

### Create Operations (7 tests)
- ✅ Create dashboard with all fields
- ✅ Create dashboard with minimal fields
- ✅ Create dashboard with null description
- ✅ Create multiple dashboards with unique names
- ✅ Reject duplicate dashboard names in same workspace
- ✅ Reject invalid dashboard name (empty, too long)
- ✅ Reject invalid config (non-object JSON)

### Read Operations (4 tests)
- ✅ Get dashboard by ID
- ✅ Get non-existent dashboard returns 404
- ✅ List dashboards with pagination
- ✅ Find dashboards by name

### Update Operations (6 tests)
- ✅ Update dashboard name and config
- ✅ Update dashboard description to null
- ✅ Update with optimistic locking (version check)
- ✅ Reject concurrent updates (version mismatch)
- ✅ Update non-existent dashboard returns 404
- ✅ Reject duplicate name on update

### Delete Operations (4 tests)
- ✅ Delete existing dashboard
- ✅ Delete non-existent dashboard returns 204
- ✅ Delete dashboard from different workspace returns 204
- ✅ Verify dashboard is actually deleted

### Slug Generation (2 tests)
- ✅ Generate URL-safe slugs from names
- ✅ Deduplicate slugs with numeric suffixes

**Run tests:**
```bash
cd apps/opik-backend
mvn clean test -Dtest=DashboardsResourceTest
```

## Documentation

No documentation updates required for this internal API. Future PRs will add:
- Frontend integration
- User-facing documentation
- API client SDK updates

andrescrz commented on apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/DashboardsResource.java:
```diff
@@ -0,0 +1,169 @@
+        service.update(id, dashboardUpdate);
+
+        // Fetch updated dashboard to return
+        Dashboard updatedDashboard = service.findById(id);
```
Better consolidate the update + refresh in a service method. Feel free to keep the granular methods there.

andrescrz commented on apps/opik-backend/src/main/java/com/comet/opik/domain/SlugUtils.java:
```diff
@@ -0,0 +1,77 @@
+ * Utility class for generating URL-safe slugs from dashboard names.
+ */
+@UtilityClass
+public class SlugUtils {
```
I'd try to use a stable and license compatible library to avoid re-inventing the while. I believe this is an option:
- https://github.com/slugify/slugify

But there would be more library options.

andrescrz commented on apps/opik-backend/src/test/java/com/comet/opik/api/resources/utils/resources/DashboardResourceClient.java:
```diff
@@ -0,0 +1,219 @@
+public class DashboardResourceClient {
+
+    private static final String RESOURCE_PATH = "%s/v1/private/workspaces/%s/dashboards";
+    private static final ObjectMapper MAPPER = new ObjectMapper();
```
Minor as in tests, but I suggest avoiding custom mappers and just go with JsonUtils.

thiagohora commented:
Thanks for the thorough review @andrescrz! I've addressed all the comments in Revision 2 (commit a629a37a0). Here's a summary of the changes:

## ✅ Addressed Comments

1. **Optimistic Locking Removed** - Removed `lastUpdatedAt` field from `DashboardUpdate` and simplified update logic to "last write wins" approach
2. **workspaceId Path Parameter** - Kept as-is for REST URL structure consistency (though not used in code since we resolve from context)
3. **Rename search → name** - Renamed query parameter with proper `@Schema` documentation clarifying it's a partial match, case insensitive
4. **Consolidate update + refresh** - Service method now returns `Dashboard` directly, eliminating the separate fetch
5. **Retry Logic Removed** - Simplified update method without retry complexity
6. **ORDER BY Optimization** - Changed from `last_updated_at DESC` to `id DESC` to leverage UUIDv7 time-based ordering with indexed primary key
7. **JsonUtils Migration** - Replaced custom `ObjectMapper` instances with `JsonUtils.getMapper()` in both `JsonNodeColumnMapper` and `DashboardResourceClient`
8. **@NonNull Annotations** - Removed unnecessary `@NonNull` from `SlugUtils` methods that already validate with `isBlank()`
9. **Unused Field Cleanup** - Removed unused `podamFactory` field from test client

## 📝 Database Index Comment

Regarding the **"Added Database Index"** suggestion for the `name` column: The unique constraint `dashboards_workspace_name_uk UNIQUE (workspace_id, name)` already creates an index on `(workspace_id, name)`, so no additional index is needed. The LIKE query for name filtering will benefit from this composite index.

## 🔄 SlugUtils Library Consideration

I kept the custom `SlugUtils` implementation for now rather than adding the `slugify` library dependency. The current implementation:
- Is simple and focused (~50 LOC)
- Has no external dependencies beyond JDK
- Handles our specific requirements (Unicode normalization, length limits, deduplication)
- Is well-tested with comprehensive test coverage

If you'd prefer to use `slugify`, I'm happy to make that change - just let me know!

All changes compile successfully and maintain backward compatibility with the existing test suite.

thiagohora commented on apps/opik-backend/src/main/java/com/comet/opik/api/resources/v1/priv/DashboardsResource.java:
```diff
@@ -0,0 +1,169 @@
+import java.net.URI;
+import java.util.UUID;
+
+@Path("/v1/private/workspaces/{workspaceId}/dashboards")
```
I missed this part thank you!

andrescrz commented on apps/opik-backend/src/main/java/com/comet/opik/domain/SlugUtils.java:
```diff
@@ -0,0 +1,76 @@
+     * @param existingCount the count of existing slugs with the same prefix
+     * @return the unique slug
+     */
+    public static String generateUniqueSlug(String baseSlug, long existingCount) {
```
Nit: I'd still check isBlank on `baseSlug`.

andrescrz reviewed (approved):
LGTM, you can move forward.

I strongly recommend changing the update method to `PATCH`. It doesn't look like a `PUT`.

My secondary commendation is using a library for generating `slugs`, but not a blocker.

The rest if minor an optional.
